### ISSUE TITLE
Improve clan balance

### ISSUE THREAD

**Issue by GoogleFrog** [ZeroK-RTS/Zero-K-Infrastructure#2949]
https://zero-k.info/Battles/Detail/1653856

![image](https://github.com/ZeroK-RTS/Zero-K-Infrastructure/assets/1318249/2b8b7b96-1543-4430-8ca7-5d535fb6089a)

Soridian and Omega[TheEnd] left this game after suggesting it be exited because they were on VC and had already played a few split games, so gave up on being balanced to the same team and did something else with their evening. I feel like the balancer could have tried harder to put clans together in the game. Perhaps the clan balance fallback should be restricted in the case of clan-pairs in a sufficiently large game. Also ideally, if clan balance is not accepted by the balancer then it could drop the clans that cause issues, rather than ignoring all clans.

**Comment by Licho1** [comment]
Does it happen more often?  Max cbalance difference is set to 70.   
By dropping clans that cause the problem you mean to try them randomly and see when balancing without them works?

**Comment by sprunk** [comment]
Yeah, let's say there's clan A (with 2 players), clan B (with 2 players), and clan C (with 3 players) in the room. Then the idealized algorithm would be:
 * try a balance where A is kept together, B is kept together, and C is kept together. If this is under cbalance threshold, use that.
 * otherwise, keep A together and B together, but split C, because they're the largest clan. If this is under cbalance threshold, use that.
 * otherwise, try (keeping A together and C together but splitting B) and (keeping B together and C together but splitting A), since that still only splits 1 clan. If the best of these two is under cbalance threshold, use that.
 * otherwise, try (keeping A together but splitting B and splitting C) and (keeping B together but splitting A and splitting C), since that keeps a small clan together. If the best of these two is under cbalance threshold, use that. If the best is under cbalance threshold, use that.
 * otherwise, try keeping C together but splitting A and splitting B. If this is under cbalance threshold, use that.
 * otherwise, split all clans (ie. use regular balance).

This is the idealized version so in practice will probably not be attainable because it grows exponentially with the number of clans. On the other hand if there are many clans then splitting the largest one should already grant good results, and keeping the smaller clans together means there's less permutations to try? Hard to tell.
